src/webgpu/shader/execution/expression/call/builtin/atomics/atomicExchange.spec.ts

```diff
@@ -18,6 +21,201 @@ fn atomicExchange(atomic_ptr: ptr<AS, atomic<T>, read_write>, v: T) -> T
 `
   )
   .params(u =>
-    u.combine('SC', ['storage', 'uniform'] as const).combine('T', ['i32', 'u32'] as const)
+    u
+      .combine('workgroupSize', workgroupSizes)
+      .combine('dispatchSize', dispatchSizes)
+      .combine('mapId', keysOf(kMapId))
+      .combine('scalarType', ['u32', 'i32'])
   )
-  .unimplemented();
+  .fn(t => {
+    const numInvocations = t.params.workgroupSize * t.params.dispatchSize;
+    const bufferNumElements = numInvocations;
+    const scalarType = t.params.scalarType;
+    const mapId = kMapId[t.params.mapId];
+    const extra = mapId.wgsl(numInvocations, t.params.scalarType); // Defines map_id()
+
+    const wgsl =
+      `
+      @group(0) @binding(0)
+      var<storage, read_write> input : array<atomic<${scalarType}>>;
+
+      @group(0) @binding(1)
+      var<storage, read_write> output : array<${scalarType}>;
+
+      @compute @workgroup_size(${t.params.workgroupSize})
+      fn main(
+          @builtin(global_invocation_id) global_invocation_id : vec3<u32>,
+          ) {
+        let id = ${scalarType}(global_invocation_id[0]);
+
+        output[id] = atomicExchange(&input[id], map_id(id * 2));
```

Comment:
Similar to `atomicCompareExchangeWeak`, the 1:1 invocation to memory location isn't testing the RMW property of `atomicExchange`. You could substitute this line with:

```ts
output[id] = atomicLoad(&input[id]);
// input[id] could change between the line above and below
atomicStore(&input[id], map_id(id * 2));
```

And still pass the test.

I suggest adding another test where:
* all the invocations `atomicExchange` to the same single memory address `A`, using the hash of the invocation IDs
* the old values returned by `atomicExchange` are atomically appended to another storage buffer `B`.
* you can then check the results by ensuring the final value held by `A` combined with the unordered values in `B` match the set of invocation ID hashes.

Reply:
Done. Added a "_basic" suffix to the original tests I wrote, and added "_advanced" ones for these you recommended.